webhook/resourcesemantics/conversion/conversion.go

```diff
@@ -0,0 +1,200 @@
+)
+
+func (r *reconciler) Convert(
+	ctx context.Context,
+	req *apixv1beta1.ConversionRequest,
+) *apixv1beta1.ConversionResponse {
+
+	if r.withContext != nil {
+		ctx = r.withContext(ctx)
+	}
+
+	res := &apixv1beta1.ConversionResponse{
+		UID: req.UID,
+		Result: metav1.Status{
+			Status: metav1.StatusSuccess,
+		},
+	}
+
+	result := make([]runtime.RawExtension, 0, len(req.Objects))
+
+	for _, obj := range req.Objects {
+		converted, err := r.convert(ctx, obj, req.DesiredAPIVersion)
+		if err != nil {
+			res.Result.Status = metav1.StatusFailure
+			res.Result.Message = err.Error()
+			break
+		}
+
+		result = append(result, converted)
+	}
+
+	res.ConvertedObjects = result
+	return res
+}
+
+func (r *reconciler) convert(
+	ctx context.Context,
+	inRaw runtime.RawExtension,
+	targetVersion string,
+) (outRaw runtime.RawExtension, err error) {
+
+	logger := logging.FromContext(ctx)
+
+	defer func() {
+		if err != nil {
+			logger.Errorf("Conversion failed: %s", err)
+		}
+	}()
+
+	inGVK, err := parseGVK(inRaw)
+	if err != nil {
+		return
+	}
+
+	logger.Infof("Converting %s to version %s", formatGVK(inGVK), targetVersion)
+
+	inGK := inGVK.GroupKind()
+	conv, ok := r.kinds[inGK]
+	if !ok {
+		err = fmt.Errorf("no conversion support for type %s", formatGK(inGVK.GroupKind()))
+		return
+	}
+
+	outGVK, err := parseAPIVersion(targetVersion, inGK.Kind)
+	if err != nil {
+		return
+	}
+
+	inZygote, ok := conv.Zygotes[inGVK.Version]
+	if !ok {
+		err = fmt.Errorf("conversion not supported for type %s", formatGVK(inGVK))
+		return
+	}
+	outZygote, ok := conv.Zygotes[outGVK.Version]
+	if !ok {
+		err = fmt.Errorf("conversion not supported for type %s", formatGVK(outGVK))
+		return
+	}
+	hubZygote, ok := conv.Zygotes[conv.HubVersion]
+	if !ok {
+		inGK := inGVK.GroupKind()
+		err = fmt.Errorf("conversion not supported for type %s", formatGK(inGK))
+		return
+	}
+
+	in := inZygote.DeepCopyObject().(ConvertibleObject)
+	hub := hubZygote.DeepCopyObject().(ConvertibleObject)
+	out := outZygote.DeepCopyObject().(ConvertibleObject)
+
+	hubGVK := inGVK.GroupKind().WithVersion(conv.HubVersion)
+	logger = logger.With(
+		zap.String("inputType", formatGVK(inGVK)),
+		zap.String("outputType", formatGVK(outGVK)),
+		zap.String("hubType", formatGVK(hubGVK)),
+	)
+
+	// TODO(dprotaso) - potentially error on unknown fields
+	if err = json.Unmarshal(inRaw.Raw, &in); err != nil {
+		err = fmt.Errorf("unable to unmarshal input: %s", err)
+		return
+	}
+
+	if inGVK.Version == conv.HubVersion {
+		hub = in
+	} else if err = hub.ConvertDown(ctx, in); err != nil {
+		err = fmt.Errorf("conversion failed to version %s for type %s -  %s", outGVK.Version, formatGVK(inGVK), err)
+		return
+	}
+
+	if outGVK.Version == conv.HubVersion {
+		out = hub
+	} else if err = hub.ConvertUp(ctx, out); err != nil {
+		err = fmt.Errorf("conversion failed to version %s for type %s -  %s", outGVK.Version, formatGVK(inGVK), err)
+		return
+	}
+
+	out.SetGroupVersionKind(outGVK)
```

Comment:
We should be applying defaults since new fields may have been added which would need to be defaulted.